File: static/solutions/cpp/java/src/main/java/com/epi/OddEven.java
package com.epi;

public class OddEven {
  //@include
  public static class OddEvenMonitor {
    public static final boolean ODD_TURN = true;
    public static final boolean EVEN_TURN = false;
    private boolean turn = ODD_TURN;

    // Need synchronized in order to call wait(), see
    // http://stackoverflow.com/questions/2779484 for discussion
    public synchronized void waitTurn(boolean oldTurn) {
      while (turn != oldTurn) {
        try {
          wait();
        } catch (InterruptedException e) {
          System.out.println("InterruptedException in wait(): " + e);
        }
      }
      // Move on, it's our turn.
    }

    // Need synchronized in order to call notify()
    public synchronized void toggleTurn() {
      turn ^= true;
      notify();
    }
  }

  public static class OddThread extends Thread {
    private final OddEvenMonitor monitor;

    public OddThread(OddEvenMonitor monitor) { this.monitor = monitor; }
    @Override
    public void run() {
      for (int i = 1; i <= 100; i += 2) {
        monitor.waitTurn(OddEvenMonitor.ODD_TURN);
        System.out.println("i = " + i);
        monitor.toggleTurn();
      }
    }
  }

  public static class EvenThread extends Thread {
    private final OddEvenMonitor monitor;

    public EvenThread(OddEvenMonitor monitor) { this.monitor = monitor; }
    @Override
    public void run() {
      for (int i = 2; i <= 100; i += 2) {
        monitor.waitTurn(OddEvenMonitor.EVEN_TURN);
        System.out.println("i = " + i);
        monitor.toggleTurn();
      }
    }
  }

  public static void main(String[] args) throws InterruptedException {
    OddEvenMonitor monitor = new OddEvenMonitor();
    Thread t1 = new OddThread(monitor);
    Thread t2 = new EvenThread(monitor);
    t1.start();
    t2.start();
    t1.join();
    t2.join();
  }
  //@exclude
}

File: solutions/cpp/odd_even.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <condition_variable>
#include <functional>
#include <iostream>
#include <mutex>
#include <thread>

using std::ref;
using std::cout;
using std::endl;
using std::thread;
using std::mutex;
using std::lock_guard;
using std::unique_lock;
using std::condition_variable;

// @include
class OddEvenMonitor {
 public:
  static const bool ODD_TURN = true;
  static const bool EVEN_TURN = false;

  OddEvenMonitor() : turn_(ODD_TURN) {}

  void WaitTurn(bool old_turn) {
    unique_lock<mutex> lock(mx_);
    while (turn_ != old_turn) {
      cond_.wait(lock);
    }
  }

  void ToggleTurn() {
    lock_guard<mutex> lock(mx_);
    turn_ = !turn_;
    cond_.notify_one();
  }

 private:
  bool turn_;
  condition_variable cond_;
  mutex mx_;
};

void OddThread(OddEvenMonitor& monitor) {
  for (int i = 1; i <= 100; i += 2) {
    monitor.WaitTurn(OddEvenMonitor::ODD_TURN);
    cout << i << endl;
    monitor.ToggleTurn();
  }
}

void EvenThread(OddEvenMonitor& monitor) {
  for (int i = 2; i <= 100; i += 2) {
    monitor.WaitTurn(OddEvenMonitor::EVEN_TURN);
    cout << i << endl;
    monitor.ToggleTurn();
  }
}
// @exclude

int main(int argc, char* argv[]) {
  OddEvenMonitor monitor;
  thread t1(OddThread, ref(monitor));
  thread t2(EvenThread, ref(monitor));
  t1.join();
  t2.join();
  return 0;
}
